tests/php/Analyzer/TupleSymbol/FnSymbolTest.php

```diff
@@ -0,0 +1,204 @@
+            )
+        );
+
+        $this->analyze($tuple);
+    }
+
+    public function providerVarNamesMustStartWithLetterOrUnderscore(): Generator
+    {
+        yield 'Start with a letter' => [
+            'paramName' => 'param-1',
+            'error' => false,
+        ];
+
+        yield 'Start with an underscore' => [
+            'paramName' => '_param-2',
+            'error' => false,
+        ];
+
+        yield 'Start with a number' => [
+            'paramName' => '1-param-3',
+            'error' => true,
+        ];
+
+        yield 'Start with an ampersand' => [
+            'paramName' => '&-param-4',
+            'error' => true,
+        ];
+
+        yield 'Start with a space' => [
+            'paramName' => ' param-5',
+            'error' => true,
+        ];
+    }
+
+    public function testOnlyOneSymbolCanFollowTheAmpersandParameter(): void
+    {
+        $this->expectException(PhelCodeException::class);
+        $this->expectExceptionMessage('Unsupported parameter form, only one symbol can follow the & parameter');
+
+        // This is the same as: (fn [& param-1 param-2])
+        $tuple = Tuple::create(
+            Symbol::create(Symbol::NAME_FN),
+            Tuple::createBracket(
+                Symbol::create('&'),
+                Symbol::create('param-1'),
+                Symbol::create('param-2'),
+            )
+        );
+
+        $this->analyze($tuple);
+    }
+
+    /** @dataProvider providerGetParams */
+    public function testGetParams(Tuple $tuple, array $expectedParams): void
+    {
+        $node = $this->analyze($tuple);
+
+        self::assertEquals($expectedParams, $node->getParams());
+    }
+
+    public function providerGetParams(): Generator
+    {
+        yield '(fn [& param-1])' => [
+            'tuple' => Tuple::create(
+                Symbol::create(Symbol::NAME_FN),
+                Tuple::createBracket(
+                    Symbol::create('&'),
+                    Symbol::create('param-1'),
+                )
+            ),
+            'expectedParams' => [
+                Symbol::create('param-1'),
+            ],
+        ];
+
+        yield '(fn [param-1 param-2 param-3])' => [
+            'tuple' => Tuple::create(
+                Symbol::create(Symbol::NAME_FN),
+                Tuple::createBracket(
+                    Symbol::create('param-1'),
+                    Symbol::create('param-2'),
+                    Symbol::create('param-3'),
+                )
+            ),
+            'expectedParams' => [
+                Symbol::create('param-1'),
+                Symbol::create('param-2'),
+                Symbol::create('param-3'),
+            ],
+        ];
+    }
+
+    /** @dataProvider providerGetBody */
+    public function testGetBody(Tuple $tuple, string $expectedBodyInstanceOf): void
+    {
+        $node = $this->analyze($tuple);
+
+        self::assertInstanceOf($expectedBodyInstanceOf, $node->getBody());
```

Comment:
@jenshaase instead of the `$expectedBodyInstanceOf`, do you know which object might be the result of each `$tuple` once analyzed (`$node->getBody()`)?

I mean, what this would be if I want to assert an `$expectedBody`
```
self::assertEquals($expectedBody, $node->getBody());
```

Reply:
`$node->getBody` returns the body of the function, wrapped in either a `DoNode` or a `LetNode`. I explained the let node in the other comment. Here is the `DoNode`. For example:
```
Tuple::create(
  Symbol::create(Symbol::NAME_FN),
  Tuple::createBracket(
    Symbol::create('x')
  ),
  Symbol::create('x'),
)
```
This is equal to `(fn [x] x)`. This will be extended by the Analyzer to `(fn [x] (do x))`.

The body of the `FnNode` is now a `DoNode`. The `DoNode` is now a little bit more complex, since `x` is resolved as local variable:
```
new DoNode(
   $nodeEnviroment, 
   array(), 
   new LocalVarNode($nodeEnviroment, $shadowVar, $location)
)
```

If you want to test the body I would recommend to use no arguments and return only literals, like numbers or strings. For example: `(fn [] 1)`.

```
Tuple::create(
  Symbol::create(Symbol::NAME_FN),
  Tuple::createBracket(
  ),
  1,
)
```

The `DoNode` should look now something like this (not tested):
```
new DoNode(
  NodeEnviroment::empty(),
  array(),
  new LiteralNode(NodeEnviroment::empty(), 1)
);
```

On last thing: The FnNode wraps everything in a `DoNode`, because it is the possible to have multiple expression. For example `(fn [] 1 2 3 4 5)` is then equal to `(fn [] (do 1 2 3 4 5))`. It is just easier to implement if the `FnNode` body is just on node instead of a array.

Reply:
Ok, I don't think it worth now to add such unit tests (for the body...). I think they are good enough for this first iteration. Maybe in a second iteration could we add such tests, but I think this PR is big and good enough as it is :)